README.md
# Python SDK

![](./images/FISCO_BCOS_Logo.svg)


[![Build Status](https://travis-ci.org/FISCO-BCOS/python-sdk.svg?branch=master)](https://travis-ci.org/FISCO-BCOS/python-sdk)
[![CodeFactor](https://www.codefactor.io/repository/github/fisco-bcos/python-sdk/badge)](https://www.codefactor.io/repository/github/fisco-bcos/python-sdk)
[![GitHub license](https://img.shields.io/github/license/FISCO-BCOS/python-sdk.svg)](https://github.com/FISCO-BCOS/python-sdk/blob/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/FISCO-BCOS/python-sdk.svg)](https://github.com/FISCO-BCOS/python-sdk/issues)
--- 

Python SDK为[FISCO BCOS](https://github.com/FISCO-BCOS/FISCO-BCOS/tree/master)提供Python API，使用FISCO BCOS Python SDK可以简单快捷的基于FISCO-BCOS进行区块链应用开发。**此版本只支持**[FISCO BCOS 2.0](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/)。


## 关键特性

- 提供调用FISCO BCOS 2.0 [JSON-RPC](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/api.html)的Python API。
- 可基于[Channel协议](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/design/protocol_description.html#channelmessage)与FISCO BCOS进行通信，保证节点与SDK安全加密通信的同时，可接收节点推送的消息。
- 支持交易解析功能：包括交易输入、交易输出、Event Log等ABI数据的拼装和解析。
- 支持合约编译，将`sol`合约编译成`abi`和`bin`文件。
- 支持基于keystore的账户管理。支持从pem文件加载ECDSA算法的私钥，以便和其他私钥管理模块互通。
- 支持本地的合约部署历史查询 (不支持链上所有部署的合约查询)。
- 支持国密(SM2,SM3,SM4算法)。如需支持国密SSL通信需要独立编译组件，参见 [cython_tassl_wrap的README](./cython_tassl_wrap)
- 支持event回调监听
- 支持liquid智能合约。需要采用[liquid开发环境](https://liquid-doc.readthedocs.io/)编译wasm合约。使用python-sdk部署和调用liquid合约时，合约名带上.wasm后缀，以和sol合约区别。


## 部署Python SDK

### 环境要求
- Python环境：python 3.6.3, 3.7.x
- FISCO BCOS节点：请参考[FISCO BCOS安装](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/installation.html#fisco-bcos)搭建

### 依赖软件

- **Ubuntu**: `sudo apt install -y zlib1g-dev libffi6 libffi-dev wget git`
- **CentOS**：`sudo yum install -y zlib-devel libffi-devel wget git`
- **MacOs**: `brew install wget npm git`

### 初始化环境(若python环境符合要求，可跳过)

#### **Linux环境初始化**

> 拉取源代码
```bash
git clone https://github.com/FISCO-BCOS/python-sdk
```

> 配置环境(安装pyenv和python，若python版本>=3.6.3可跳过本步)
```bash
# 获取python版本
python --version

## --------若python版本小于3.6.3，执行下面流程--------------------------------------
# 判断python版本，并为不符合条件的python环境安装python 3.7.3的虚拟环境，命名为python-sdk
# 若python环境符合要求，可以跳过此步
# 若脚本执行出错，请检查是否参考[依赖软件]说明安装了依赖
# 提示：安装python-3.7.3可能耗时比较久
cd python-sdk && bash init_env.sh -p

## --------若通过bash init_env.sh -p安装了python-sdk虚拟环境，执行下面流程-------------
# 激活python-sdk虚拟环境
source ~/.bashrc && pyenv activate python-sdk && pip install --upgrade pip
```

#### **Windows环境初始化**

在Windows运行Python SDK，需要按照以下步骤安装依赖软件并配置合约编译器：

**安装依赖软件**

- 直接安装[Python-3.7.x](https://www.python.org/downloads/release/python-373/)和[git](https://git-scm.com/download/win)软件  
> python环境变量配置可参考[这里](https://jingyan.baidu.com/article/b0b63dbff271e24a4830708d.html)

- [Visual C++ 14.0库](https://visualstudio.microsoft.com/downloads)
> (注：Microsoft Visual C++ 14.0 is required. Get it with "Microsoft Visual C++ Build Tools"解决方法: https://visualstudio.microsoft.com/downloads （注意选择vs 2005即14.0版）或 https://pan.baidu.com/s/1ZmDUGZjZNgFJ8D14zBu9og 提取码: zrby)

- 下载Windows版本solc, 点击[这里](https://github.com/ethereum/solidity/releases/download/v0.4.25/solidity-windows.zip)下载
> solc编译器下载成功后，解压，将其中的 solc.exe 文件复制 ${python-sdk}\bin 目录下。若 python-sdk 路为 D:\\open-source\\python-sdk, 则 solc.exe 文件复制路径为D:\\open-source\\python-sdk\\bin\\solc.exe 

**拉取源代码**

打开 git，在任意目录执行如下命令
```bash
git clone https://github.com/FISCO-BCOS/python-sdk
```

**配置solc编译器路径**
```bash
# 修改client_config.py.template: 
# 配置solc编译器路径，若solc存放路径为D:\\open-source\\python-sdk\\bin\\solc.exe，则solc_path配置如下：
solc_path = "D:\\open-source\\python-sdk\\bin\\solc.exe"

# 将client_config.py.template拷贝到client_config.py
```

### **安装Python SDK依赖**

```bash
cd python-sdk 
pip install -r requirements.txt
```

> **若因网络原因，安装依赖失败，可使用清华的pip源下载，安装命令如下：**

```bash
pip install -i https://pypi.tuna.tsinghua.edu.cn/simple -r requirements.txt
```

### 初始化配置(Windows环境可跳过)

```bash
# 该脚本执行操作如下：
# 1. 拷贝client_config.py.template -> client_config.py
# 2. 安装solc编译器
bash init_env.sh -i
```

> **若MacOS环境solc安装较慢，可在python-sdk目录下执行如下命令安装solcjs**，python-sdk自动加载nodejs编译器：

```bash
# 安装编译器
npm install solc@v0.4.25
```

> 若没有执行以上初始化步骤，需要将`contracts/`目录下的`sol`代码手动编译成`bin`和`abi`文件并放置于`contracts`目录，才可以部署和调用相应合约。合约编译可以使用[remix](https://remix.ethereum.org)


## 配置Channel通信协议

Python SDK支持使用[Channel协议](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/design/protocol_description.html#channelmessage-v1)与FISCO BCOS节点通信，通过SSL加密通信保障SDK与节点通信的机密性。

设SDK连接的节点部署在目录`~/fisco/nodes/127.0.0.1`目录下，则通过如下步骤使用Channel协议：

**配置Channel信息**

在节点目录下的 config.ini 文件中获取 channel_listen_port, 这里为20200  
```bash
[rpc]
    listen_ip=0.0.0.0
    channel_listen_port=20200
    jsonrpc_listen_port=8545
```
    
切换到python-sdk目录，修改 client_config.py 文件中`channel_host`为实际的IP，`channel_port`为上步获取的`channel_listen_port`：

```bash
channel_host = "127.0.0.1"
channel_port = 20200
```

**配置证书**

```bash
# 若节点与python-sdk位于不同机器，请将节点sdk目录下所有相关文件拷贝到bin目录
# 若节点与sdk位于相同机器，直接拷贝节点证书到SDK配置目录
cp ~/fisco/nodes/127.0.0.1/sdk/* bin/
```

**配置证书路径**

  - `client_config.py`的`channel_node_cert`和`channel_node_key`选项分别用于配置SDK证书和私钥
  - `release-2.1.0`版本开始，SDK证书和私钥更新为`sdk.crt`和`sdk.key`，配置证书路径前，请先检查上步拷贝的证书名和私钥名，并将`channel_node_cert`配置为SDK证书路径，将`channel_node_key`配置为SDK私钥路径

检查从节点拷贝的sdk证书路径，若sdk证书和私钥路径分别为`bin/sdk.crt`和`bin/sdk.key`，则`client_config.py`中相关配置项如下：

```bash
channel_node_cert = "bin/sdk.crt"  # 采用channel协议时，需要设置sdk证书,如采用rpc协议通信，这里可以留空
channel_node_key = "bin/sdk.key"   # 采用channel协议时，需要设置sdk私钥,如采用rpc协议通信，这里可以留空
```

若sdk证书和私钥路径分别为`bin/node.crt`和`bin/node.key`，则`client_config.py`中相关配置项如下:
```bash
channel_node_cert = "bin/node.crt"  # 采用channel协议时，需要设置sdk证书,如采用rpc协议通信，这里可以留空
channel_node_key = "bin/node.key"   # 采用channel协议时，需要设置sdk私钥,如采用rpc协议通信，这里可以留空
```

**国密支持**
 -  支持国密版本的非对称加密、签名验签(SM2), HASH算法(SM3),对称加解密(SM4)
 -  国密版本在使用上和非国密版本基本一致，主要是配置差异。
 -  国密版本sdk同一套代码可以连接国密和非国密的节点，需要根据不同的节点配置相应的IP端口
 -  截止2021.02版本，连接国密节点需使用非国密节点和SDK证书（节点生成非国密连接证书，参见节点的build_chain.sh和企业版搭链等文档），TLS国密认证实现中
 -  因为当前版本的实现里，账户文件格式有差异，所以国密的账户文件和ECDSA的账户文件采用不同的配置
 -  国密SSL目前需要手动编译，配置，方法参见 [cython_tassl_wrap的README](./cython_tassl_wrap)

连接国密节点时，有以下相关的配置项需要修改和确认，IP端口也需要确认是指向国密版本节点
```bash
crypto_type = "GM" 	#密码算法选择: 大小写不敏感："GM" 标识国密, "ECDSA" 或其他是椭圆曲线默认实现。
ssl_type = "GM" # 和节点tls通信方式，如设为gm，则使用国密证书认证和加密
gm_account_keyfile = "gm_account.json"  #国密账号的存储文件，可以加密存储,如果留空则不加载
gm_account_password = "123456" 		#如果不设密码，置为None或""则不加密
gm_solc_path = "./bin/solc/v0.4.25/solc-gm" #合约编译器配置，通过执行bash init_env.sh -i命令下载
以及5个证书配置
```


**使用Channel协议访问节点**

```bash
# 获取FISCO BCOS节点版本号
./console.py getNodeVersion
```

**Event事件回调**
 -  针对已经部署在链上某个地址的合约，先注册要监听的事件，当合约被交易调用，且生成事件时，节点可以向客户端推送相应的事件
 -  事件定义如有indexed类型的输入，可以指定监听某个特定值作为过滤，如事件定义为 on_set(string name,int indexed value),可以增加一个针对value的topic监听，只监听value=5的事件
 -  具体实现参考demo_event_callback.py,使用的命令行为：
```bash
params: contractname address event_name indexed
        1. contractname :       合约的文件名,不需要带sol后缀,默认在当前目录的contracts目录下
        2. address :    十六进制的合约地址,或者可以为:last,表示采用bin/contract.ini里的记录
        3. event_name : 可选,如不设置监听所有事件
        4. indexed :    可选,根据event定义里的indexed字段,作为过滤条件)

        eg: for contract sample [contracts/HelloEvent.sol], use cmdline:

        python demo_event_callback.py HelloEvent last
        --listen all event at all indexed ：

        python demo_event_callback.py HelloEvent last on_set
        --listen event on_set(string newname) （no indexed）：

        python demo_event_callback.py HelloEvent last on_number 5
        --listen event on_number(string name,int indexed age), age ONLY  5 ：

```


## SDK使用示例

**查看SDK使用方法**

> **windows环境下执行console.py请使用`.\console.py`或者`python console.py`**

```bash
# 查看SDK使用方法
./console.py usage

# 获取节点版本
./console.py getNodeVersion
```

**部署HelloWorld合约**
```bash
$ ./console.py deploy HelloWorld  

INFO >> user input : ['deploy', 'HelloWorld']

backup [contracts/HelloWorld.abi] to [contracts/HelloWorld.abi.20190807102912]
backup [contracts/HelloWorld.bin] to [contracts/HelloWorld.bin.20190807102912]
INFO >> compile with solc compiler
deploy result  for [HelloWorld] is:
 {
    "blockHash": "0x3912605dde5f7358fee40a85a8b97ba6493848eae7766a8c317beecafb2e279d",
    "blockNumber": "0x1",
    "contractAddress": "0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce",
    "from": "0x95198b93705e394a916579e048c8a32ddfb900f7",
    "gasUsed": "0x44ab3",
    "input": "0x6080604052...省略若干行...c6f2c20576f726c642100000000000000000000000000",
    "logs": [],
    "logsBloom": "0x000...省略若干行...0000",
    "output": "0x",
    "status": "0x0",
    "to": "0x0000000000000000000000000000000000000000",
    "transactionHash": "0xb291e9ca38b53c897340256b851764fa68a86f2a53cb14b2ecdcc332e850bb91",
    "transactionIndex": "0x0"
}
on block : 1,address: 0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce 
address save to file:  bin/contract.ini
```

**调用HelloWorld合约**

```bash
# 合约地址：0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce
# 调用接口：get
$./console.py  call HelloWorld 0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce get 

INFO >> user input : ['call', 'HelloWorld', '0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce', 'get']

INFO >> call HelloWorld , address: 0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce, func: get, args:[]
INFO >> call result: ('Hello, World!',)

# 合约名：HelloWorld
# 合约地址：0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce
# 调用接口：set
# 参数："Hello, FISCO"
$ ./console.py sendtx HelloWorld 0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce set "Hello, FISCO"

INFO >> user input : ['sendtx', 'HelloWorld', '0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce', 'set', 'Hello, FISCO']

INFO >> sendtx HelloWorld , address: 0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce, func: set, args:['Hello, FISCO']

INFO >>  receipt logs : 
INFO >> transaction hash :  0xc20cbc6b0f28ad8fe1c560c8ce28c0e7eb7719a4a618a81604ac87ac46cc60f0
tx input data detail:
 {'name': 'set', 'args': ('Hello, FISCO',), 'signature': 'set(string)'}
receipt output : ()

# 调用get接口获取更新后字符串
$./console.py call HelloWorld 0x2d1c577e41809453c50e7e5c3f57d06f3cdd90ce get

INFO >> user input : ['call', 'HelloWorld', '0x42883e01ac97a3a5ef8a70c290abe0f67913964e', 'get']


Changelog.md
### v0.9.3
(2021-03-30)

**更新**
- 重构`console.py`
- 控制台`console.py`支持数组类型的参数输入


### v0.9.2
(2021-03-30)

**新特性**
- 支持国密算法
- 提供合约事件回调接口

**更新**

- 支持使用`console.py`部署和调用参数为`bytes`的合约

**修复**

- 修复解析Precompiled合约回执前没有判断回执状态导致的解析异常


### v0.9.1
(2020-06-19)

**更新**
- 支持带参数的合约部署
- 优化交易回执处理流程
- 添加客户端请求超时逻辑


### v0.9.0

(2019-08-16)

**新特性** 

- 提供调用FISCO BCOS 2.0 JSON-RPC的Python API。
- 可基于Channel协议与FISCO BCOS进行通信，保证节点与SDK安全加密通信的同时，接收节点推送的消息。
- 支持交易解析功能：包括交易输入、交易输出、Event Log等ABI数据的拼装和解析。
- 支持合约编译，将sol合约编译成abi和bin文件。
- 支持基于keystore的账户管理。


ci/ci_check.sh
#!/bin/bash
python_bin="python"
LOG_ERROR()
{
    local content=${1}
    echo -e "\033[31m[ERROR] ${content}\033[0m"
    exit 1
}

execute_cmd()
{
    command="${1}"
    #LOG_INFO "RUN: ${command}"
    eval ${command}
    ret=$?
    if [ $ret -ne 0 ];then
        LOG_ERROR "FAILED execution of command: ${command}"
	if [ -d "nodes" ];then
	   bash nodes/127.0.0.1/stop_all.sh && rm -rf nodes
        fi
    fi
}

LOG_INFO()
{
    local content=${1}
    echo -e "\033[32m[INFO] ${content}\033[0m"
}

cur_path=$(execute_cmd "pwd")

# build blockchain
function build_blockchain()
{
  if [ -f "nodes/127.0.0.1/stop_all.sh" ];then
    execute_cmd "bash nodes/127.0.0.1/stop_all.sh"
  fi
  execute_cmd "rm -rf nodes"
  # download build_chain.sh
  tag=$(curl -sS "https://gitee.com/api/v5/repos/FISCO-BCOS/FISCO-BCOS/tags" | grep -oe "\"name\":\"v[2-9]*\.[0-9]*\.[0-9]*\"" | cut -d \" -f 4 | sort -V | tail -n 1)
  LOG_INFO "--- current tag: $tag"
  curl -LO "https://github.com/FISCO-BCOS/FISCO-BCOS/releases/download/${tag}/build_chain.sh" && chmod u+x build_chain.sh
  # get_buildchain.sh may fail due to access github api failed
  #bash <(curl -s https://raw.githubusercontent.com/FISCO-BCOS/FISCO-BCOS/dev/tools/get_buildchain.sh)
  
  if [ ! -f "build_chain.sh" ];then
     LOG_ERROR "get build_chain.sh failed!"
  fi
  execute_cmd "chmod a+x build_chain.sh"
  # build the blockchain
  ./build_chain.sh -l "127.0.0.1:4"
  # copy certificate
  execute_cmd "cp nodes/127.0.0.1/sdk/* bin/"
  execute_cmd "cp bin/node.crt bin/sdk.crt"
  execute_cmd "cp bin/node.key bin/sdk.key"
}

# start the nodes
function start_nodes()
{
   execute_cmd "./nodes/127.0.0.1/start_all.sh"
}

# stop the nodes
function stop_nodes()
{
   execute_cmd "./nodes/127.0.0.1/stop_all.sh"
}

function getBlockNumber()
{
    execute_cmd "${python_bin} console.py getBlockNumber | grep -v INFO |grep -v BcosClient | grep -v crypto | awk -F':' '{print \$2}' | awk '\$1=\$1'"
}

# test the common jsonRPC interface
function test_common_rpcInterface()
{
   LOG_INFO "## test commonRPCInterface..."
   # getNodeVersion
   execute_cmd "${python_bin} console.py getNodeVersion"
   # usage
   execute_cmd "${python_bin} console.py usage"
   # list
   execute_cmd "${python_bin} console.py list"
   # demo_get
   execute_cmd "${python_bin} demo_get.py"
   LOG_INFO "## test commonRPCInterface finished..."
} 

# test the contract
function test_contract()
{
    LOG_INFO "## test contract..."
    # deploy contract with params
    local contract_addr=$(execute_cmd "${python_bin} console.py deploy HelloWorldTest testCase | grep "on.*block.*address:" | awk -F':' '{print \$3}' | awk '\$1=\$1'")
    echo "#### contract_addr is: ${contract_addr}"

    local ret=$(execute_cmd "${python_bin} console.py call HelloWorldTest ${contract_addr} get |  grep testCase")
    if [ "$ret" == "" ];then
	LOG_ERROR "deploy contract HelloWorldTest with params failed, ret: ${ret}"
    fi

    init_blockNumber=$(getBlockNumber)
    # deploy and get contract address
    contract_addr=$(execute_cmd "${python_bin} console.py deploy HelloWorld | grep "on.*block.*address:" | awk -F':' '{print \$3}' | awk '\$1=\$1'")
    echo "#### contract_addr is: ${contract_addr}"

    updated_blockNumber=$(getBlockNumber)
    if [ $(($init_blockNumber + 1)) -ne $((updated_blockNumber)) ];then
        LOG_ERROR "deploy contract failed for blockNumber hasn't increased"
    fi
    execute_cmd "${python_bin} console.py getCode ${contract_addr}"
    # test cns precompile
    LOG_INFO "## test CNS precompile"
    LOG_INFO "registerCNS..."
    version="1.0_${updated_blockNumber}"
    execute_cmd "${python_bin} console.py registerCNS HelloWorld \$contract_addr "\${version}"" 
    LOG_INFO "queryCNSByName..."
    query_addr=$(execute_cmd "${python_bin} console.py queryCNSByName HelloWorld | grep "ContractAddress"|  tail -n 1 | awk -F':' '{print \$2}' | awk '\$1=\$1'")
    
    if [ $(echo ${query_addr} | tr 'a-z' 'A-Z') != $(echo ${contract_addr} | tr 'a-z' 'A-Z') ];then
        LOG_ERROR "queryCNSByName failed for inconsistent contract address"
    fi
    LOG_INFO "queryCNSByNameAndVersion..."
    query_addr=$(execute_cmd "${python_bin} console.py queryCNSByNameAndVersion HelloWorld \"\$version\" | grep "ContractAddress"|  awk -F':' '{print \$2}' | awk '\$1=\$1'")
    if [ $(echo ${query_addr} | tr 'a-z' 'A-Z') != $(echo ${contract_addr} | tr 'a-z' 'A-Z') ];then
        LOG_ERROR "queryCNSByName failed for inconsistent contract addresss"
    fi
    # test getBlockByNumber
    execute_cmd "${python_bin} console.py getBlockByNumber \$((\$updated_blockNumber))"
    # test call HelloWord
    local ret=$(execute_cmd "${python_bin} console.py call HelloWorld \${contract_addr} \"get\"| grep "Hello" ")
    # check call result
    if [ -z "${ret}" ];then
        LOG_ERROR "call HelloWorld failed!"
    fi
    # test sendtx
    execute_cmd "${python_bin} console.py sendtx HelloWorld  \${contract_addr} \"set\" \"Hello,FISCO\""
    # check call result
    local ret=$(execute_cmd "${python_bin} console.py call HelloWorld \${contract_addr} \"get\"| grep "Hello,FISCO"")
    # check result
    if [ -z "${ret}" ];then
        LOG_ERROR "sendtx failed to set HelloWorld failed!"
    fi
    # deploy TableTest
    execute_cmd "${python_bin} console.py deploy TableTest"
    execute_cmd "${python_bin} console.py deploy ParallelOk"
    LOG_INFO "## test contract finished..."
}

# test account
function test_account()
{
    LOG_INFO "## test account..."
    local file_path="bin/accounts/test_account.keystore"
    execute_cmd "rm -rf \${file_path}"
    # new account
    LOG_INFO ">> test newaccount..."
    local addr=$(execute_cmd "${python_bin} console.py newaccount test_account "123456" | grep "address" | grep -v "new" | awk -F':' '{print \$2}' | awk '\$1=\$1'")
    if [ ! -f "${file_path}" ];then
		# LOG_ERROR "new account failed!"
		local file_path="bin/accounts/test_account.json"
		if [ ! -f "${file_path}" ];then
			LOG_ERROR "new account failed!"
		fi
	fi

    # show account
    LOG_INFO ">> test showaccount..."
    local addr2=$(execute_cmd "${python_bin} console.py showaccount test_account \"123456\" | grep "address" | grep -v "new" | awk -F':' '{print \$2}' | awk '\$1=\$1'")
    
    # check address
    if [ "${addr}" != "${addr2}" ];then
        LOG_ERROR "check newaccount/showaccount failed for inconsistent store/load address, store_addr: ${addr}, load_addr: ${addr2}"
    fi
    LOG_INFO "## test account finished..."
}
# test permission precompile
function test_permission_precompile()
{
    account="0x95198B93705e394a916579e048c8A32DdFB900f7"
    table="t_test"
    # create table
    execute_cmd "${python_bin} console.py createTable ${table} t_test test test"
    # query table
    execute_cmd "${python_bin} console.py desc ${table}"
    # grantPermissionManager
    execute_cmd "${python_bin} console.py grantPermissionManager ${account}"
    # listPermissionManager
    execute_cmd "${python_bin} console.py listPermissionManager | grep -i ${account}"
    granted_account="0xcDF16CeF9004b1ECCf464Ae559996712E250D5A9"
     # grantNodeManager
    execute_cmd "${python_bin} console.py grantNodeManager ${granted_account}"
    # listNodeManager
    execute_cmd "${python_bin} console.py listNodeManager | grep -i ${granted_account}"
    # grantCNSManager
    execute_cmd "${python_bin} console.py grantCNSManager ${granted_account}"
    # listCNSManager
    execute_cmd "${python_bin} console.py listCNSManager | grep -i ${granted_account}"
    # grantSysConfigManager
    execute_cmd "${python_bin} console.py grantSysConfigManager ${granted_account}"
    # listSysConfigManager
    execute_cmd "${python_bin} console.py listSysConfigManager | grep -i ${granted_account}"
    # grantUserTableManager
    execute_cmd "${python_bin} console.py grantUserTableManager ${table} ${granted_account}"
    # listUserTableManager
    execute_cmd "${python_bin} console.py listUserTableManager ${table} |grep -i ${granted_account}"
    # grantDeployAndCreateManager
    execute_cmd "${python_bin} console.py grantDeployAndCreateManager ${granted_account}"
    # listDeployAndCreateManager
    execute_cmd "${python_bin} console.py listDeployAndCreateManager | grep -i ${granted_account}"

    # call revoke
    # revokeUserTableManager
    execute_cmd "${python_bin} console.py revokeUserTableManager ${table} ${granted_account}"
    # revokeDeployAndCreateManager
    execute_cmd "${python_bin} console.py revokeDeployAndCreateManager ${granted_account}"
    # revokeNodeManager
    execute_cmd "${python_bin} console.py revokeNodeManager ${granted_account}"
    # revokeCNSManager
    execute_cmd "${python_bin} console.py revokeCNSManager ${granted_account}"
    # revokeSysConfigManager
    execute_cmd "${python_bin} console.py revokeSysConfigManager ${granted_account}"
    # revokePermissionManager
    execute_cmd "${python_bin} console.py revokePermissionManager ${granted_account}"

    # call list again
    ${python_bin} console.py listUserTableManager ${table}| grep -i ${granted_account}
    command_list="listDeployAndCreateManager listNodeManager listCNSManager listSysConfigManager listPermissionManager"
    for command in ${command_list};do
        ${python_bin} console.py ${command} | grep -i ${granted_account}
    done
}

# test rpc interfaces
function test_rpc_command()
{
    # execute command with empty params
    empty_cmd_list="getNodeVersion getBlockNumber getPbftView \
    getSealerList getObserverList getConsensusStatus getSyncStatus \
    getPeers getGroupPeers getNodeIDList getGroupList getPendingTxSize \
    getTotalTransactionCount getPendingTransactions"
    for cmd in ${empty_cmd_list};do
        execute_cmd "${python_bin} console.py ${cmd}"
    done
    # execute command with one param
    one_param="getBlockHashByNumber"
    for cmd in ${one_param};do
        execute_cmd "${python_bin} console.py ${cmd} 0"
    done 
    blockHash=$(${python_bin} console.py getTransactionByBlockNumberAndIndex 1 0 | grep "blockHash" | awk -F':' '{print $2}'| awk '$1=$1' | cut -d'"' -f2)
    txHash=$(${python_bin} console.py getTransactionByBlockNumberAndIndex 1 0 | grep "hash" |grep -v block | awk -F':' '{print $2}'| awk '$1=$1' | cut -d'"' -f2)
    # getBlockByHash
    execute_cmd "${python_bin} console.py getBlockByHash ${blockHash} True"
    execute_cmd "${python_bin} console.py getBlockByHash ${blockHash} False"
    execute_cmd "${python_bin} console.py getBlockByHash ${blockHash}"
    execute_cmd "${python_bin} console.py getBlockByNumber 1 True"
    execute_cmd "${python_bin} console.py getBlockByNumber 1 False"
    execute_cmd "${python_bin} console.py getBlockByNumber 1"
    # getTransactionByHash
    execute_cmd "${python_bin} console.py getTransactionByHash ${txHash} HelloWorld"
    execute_cmd "${python_bin} console.py getTransactionByHash ${txHash}"
    # getTransactionReceipt
    execute_cmd "${python_bin} console.py getTransactionReceipt ${txHash} HelloWorld"
    execute_cmd "${python_bin} console.py getTransactionReceipt ${txHash}"
    # getTransactionByBlockHashAndIndex
    execute_cmd "${python_bin} console.py getTransactionByBlockHashAndIndex ${blockHash} 0 HelloWorld"
    execute_cmd "${python_bin} console.py getTransactionByBlockHashAndIndex ${blockHash} 0"
}


# test consensus precompile
function test_consensus_precompile()
{
    # get node Id for node1
    node_id=$(execute_cmd "cat nodes/127.0.0.1/node1/conf/node.nodeid")
    # test getSealerList
    LOG_INFO "getSealerList..."
    obtained_node_id=$(execute_cmd "${python_bin} console.py "getSealerList" | grep ${node_id}")
    if [ -z "${obtained_node_id}" ];then
        LOG_ERROR "getSealerList failed for without node1: ${node_id}"
    fi
    # test removeNode
    LOG_INFO "removeNode..."
    execute_cmd "${python_bin} console.py \"removeNode\" \${node_id}"
    obtained_node_id=$(${python_bin} console.py "getSealerList" | grep ${node_id})
    sleep 1s
    if [ ! -z "${obtained_node_id}" ];then
        LOG_ERROR "remove node1: ${node_id} failed"
    fi
    obtained_node_id=$(${python_bin} console.py "getObserverList" | grep ${node_id})
    if [ ! -z "${obtained_node_id}" ];then
        LOG_ERROR "remove node1: ${node_id} failed"
    fi
    # test addObserver
    LOG_INFO "addObserver..."
    execute_cmd "${python_bin} console.py \"addObserver\" \${node_id}"
    sleep 1s
    obtained_node_id=$(${python_bin} console.py "getSealerList" | grep ${node_id})
    if [ ! -z "${obtained_node_id}" ];then
        LOG_ERROR "addObserver for ${node_id} failed for search in sealer list"
    fi
    obtained_node_id=$(${python_bin} console.py "getObserverList" | grep ${node_id})
    if [ -z "${obtained_node_id}" ];then
        LOG_ERROR "addObserver for: ${node_id} failed"
    fi
    # test addSealer
    execute_cmd "${python_bin} console.py \"addSealer\" \${node_id}"
    sleep 1s
    obtained_node_id=$(${python_bin} console.py "getSealerList" | grep ${node_id})
    if [ -z "${obtained_node_id}" ];then
        LOG_ERROR "addObserver for ${node_id} failed"
    fi
    sleep 1s
    obtained_node_id=$(${python_bin} console.py "getObserverList" | grep ${node_id})
    if [ ! -z "${obtained_node_id}" ];then
        LOG_ERROR "addObserver for: ${node_id} failed  for search in observer list"
    fi
}

function get_config_by_key()
{
    key="${1}"
    value=$(execute_cmd "${python_bin} console.py \"getSystemConfigByKey\" \${key} | grep -v INFO | awk -F':' '{print \$2}' | awk '\$1=\$1'")
    echo "${value}"
}

function set_config_by_key()
{
    key="${1}"
    value="${2}"
    execute_cmd "${python_bin} console.py \"setSystemConfigByKey\" \${key} \${value}"
}

# test sys_config precompile
function test_sys_config()
{
    LOG_INFO "test_sys_config, key: tx_count_limit..."
    tx_limit=$(get_config_by_key "tx_count_limit")
    if [ "${tx_limit}" != "1000" ];then
        LOG_ERROR "invalid tx_count_limit, should be 1000"
    fi
    updated_tx_limit="500"
    set_config_by_key "tx_count_limit" ${updated_tx_limit}
    tx_limit=$(get_config_by_key "tx_count_limit")
    if [ "${tx_limit}" != "${updated_tx_limit}" ];then
        LOG_ERROR "set tx_count_limit failed, should be ${updated_tx_limit}"
    fi

    set_config_by_key "tx_count_limit" "1000"
    tx_limit=$(get_config_by_key "tx_count_limit")
    if [ "${tx_limit}" != "1000" ];then
        LOG_ERROR "set tx_count_limit failed, should be 1000"
    fi
    LOG_INFO "test_sys_config, key: tx_count_limit Succ"

    # test tx_gas_limit
    LOG_INFO "test_sys_config, key: tx_gas_limit..."
    tx_gas_limit=$(get_config_by_key "tx_gas_limit")
    tx_gas="300000000"
    if [ "${tx_gas_limit}" != "${tx_gas}" ];then
        LOG_ERROR "get tx_gas_limit failed, should be ${tx_gas}"
    fi
    # set tx_gas_limit
    updated_gas="500000000"
    set_config_by_key "tx_gas_limit" "${updated_gas}"
    tx_gas_limit=$(get_config_by_key "tx_gas_limit")
    if [ "${tx_gas_limit}" != "${updated_gas}" ];then
        LOG_ERROR "set tx_gas_limit failed, should be ${updated_gas}"
    fi

    # set tx_gas_limit to tx_gas
    set_config_by_key "tx_gas_limit" "${tx_gas}"
    tx_gas_limit=$(get_config_by_key "tx_gas_limit")
    if [ "${tx_gas_limit}" != "${tx_gas}" ];then
        LOG_ERROR "set tx_gas_limit failed, should be ${tx_gas}"
    fi
}

function test_precompile()
{
    test_consensus_precompile
    test_sys_config
    #test_permission_precompile
}

# test_rpc
function test_rpc()
{
    test_common_rpcInterface
    test_contract
    test_account
    test_rpc_command
}
# test channel
function test_channel()
{
    # update config to channel
    sed -i "s/client_protocol = \"rpc\"/client_protocol = \"channel\"/g" client_config.py
    test_rpc
    test_precompile
}

# test rpc
function modify_to_rpc()
{
    # update config to channel
    sed -i "s/client_protocol = \"channel\"/client_protocol = \"rpc\"/g" client_config.py
}

function main()
{
   execute_cmd "cp client_config.py.template client_config.py"
   # test rpc
   modify_to_rpc   
   build_blockchain
   start_nodes
   # callback demo_transaction
   execute_cmd "${python_bin} demo_transaction.py"
   test_rpc
   test_precompile
   test_channel
   stop_nodes
}
main
